server/src/main/java/org/elasticsearch/ingest/IngestCtxMap.java

```diff
@@ -43,7 +45,7 @@ final class IngestCtxMap extends CtxMap<IngestDocMetadata> {
         ZonedDateTime timestamp,
         Map<String, Object> source
     ) {
-        super(new HashMap<>(source), new IngestDocMetadata(index, id, version, routing, versionType, timestamp));
+        super(source, new IngestDocMetadata(index, id, version, routing, versionType, timestamp));
```

Comment:
There's no way this can burn us in any existing scripts, is there? I don't think they can ever call this constructor, so it couldn't be a problem. Just trying to think through potential problems.

Reply:
It's the `ctx` of a `script` processor, so by the time the script processor has a handle on it, it would have already been duplicated off the once (at the very very beginning new `IngestService.newIngestDocument`). It's not something you can call from a `script` processor.

I spent a fair amount of time reading through the various constructors and copy constructors of the `IngestDocument` and its friends, and I really do think I've covered all the bases.

Reply:
But philosophically, if it's a bug to *not* rehash, then what about the bug of re-hashing but **only one layer deep** -- these maps are almost always deeply nested, if this code had a job to do then it never actually did it.